core/src/test/java/org/apache/calcite/avatica/AvaticaResultSetThrowsSqlExceptionTest.java

```diff
@@ -75,11 +90,118 @@ public void testUpdateNull() throws SQLException {
     final TestDriver driver = new TestDriver();
     try (Connection connection = driver.connect("jdbc:test", properties);
          ResultSet resultSet =
-             connection.createStatement().executeQuery("SELECT * FROM TABLE")) {
+                 connection.createStatement().executeQuery("SELECT * FROM TABLE")) {
       thrown.expect(SQLFeatureNotSupportedException.class);
       resultSet.updateNull(1);
     }
   }
-}
 
+  @Test
+  public void testCommonCursorStates() throws SQLException {
+    final ResultSet resultSet = getResultSet();
+
+    // right after statement execution, result set is before first row
+    assert resultSet.isBeforeFirst();
+
+    // retrieve each row until the last one
+    while (!resultSet.isAfterLast()) {
```

Comment:
Infinite loop is possible. Please add a counter

Reply:
I have added a limit of at most 2 rows, that is sufficient for the test at hand. 
Do you see any better way to enforce a reasonable limit?
There is no way to know beforehand the size of the result set, you can just check the return value of next and isAfterLast (this last was my previous attempt).